File: Practical_04.py
# Function to perform Selection Sort
def selection_sort(arr):
    n = len(arr)
    for i in range(n):
        min_index = i
        for j in range(i + 1, n):
            if arr[j] < arr[min_index]:
                min_index = j
        # Swap the found minimum element with the first element
        arr[i], arr[min_index] = arr[min_index], arr[i]

# Function to perform Bubble Sort
def bubble_sort(arr):
    n = len(arr)
    for i in range(n):
        for j in range(0, n-i-1):
            if arr[j] > arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j]

# Function to display the top 5 scores
def display_top_five(arr):
    top_five = arr[-5:] if len(arr) >= 5 else arr
    print("\nTop Five Scores:")
    for i, score in enumerate(reversed(top_five), start=1):
        print(f"{i}. {score:.2f}%")

# Main function
def main():
    # Input: Number of students and their first-year percentages
    n = int(input("Enter the number of students: "))
    percentages = []

    for i in range(n):
        score = float(input(f"Enter the percentage of student {i+1}: "))
        percentages.append(score)

    # Sorting and displaying top five scores using Selection Sort
    selection_sorted_percentages = percentages.copy()
    selection_sort(selection_sorted_percentages)
    print("\nScores sorted using Selection Sort:")
    print(selection_sorted_percentages)
    display_top_five(selection_sorted_percentages)

    # Sorting and displaying top five scores using Bubble Sort
    bubble_sorted_percentages = percentages.copy()
    bubble_sort(bubble_sorted_percentages)
    print("\nScores sorted using Bubble Sort:")
    print(bubble_sorted_percentages)
    display_top_five(bubble_sorted_percentages)

# Run the program
if __name__ == "__main__":
    main()

File: Practical_04.c
#include <stdio.h>

#define MAX_STUDENTS 100

// Function for selection sort
void selectionSort(float arr[], int n) {
    int i, j, minIndex;
    float temp;
    
    for (i = 0; i < n - 1; i++) {
        minIndex = i;
        
        // Find the minimum element in unsorted array
        for (j = i + 1; j < n; j++) {
            if (arr[j] < arr[minIndex]) {
                minIndex = j;
            }
        }
        
        // Swap the found minimum element with the first element
        if (minIndex != i) {
            temp = arr[i];
            arr[i] = arr[minIndex];
            arr[minIndex] = temp;
        }
    }
}

// Function for bubble sort
void bubbleSort(float arr[], int n) {
    int i, j;
    float temp;
    
    for (i = 0; i < n - 1; i++) {
        // Flag to check if any swapping is done in this pass
        int swapped = 0;
        
        for (j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                // Swap if the element is greater than the next one
                temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
                swapped = 1;
            }
        }
        
        // If no elements were swapped, break out of the loop
        if (!swapped) {
            break;
        }
    }
}

// Function to display the top five scores
void displayTopFiveScores(float arr[], int n) {
    printf("Top 5 scores:\n");
    
    int topCount = (n < 5) ? n : 5;  // Display only top 5 or less if less than 5 students
    
    for (int i = n - 1; i >= n - topCount; i--) {
        printf("%.2f\n", arr[i]);
    }
}

int main() {
    int n;
    float scores[MAX_STUDENTS];

    // Input number of students
    printf("Enter number of students: ");
    scanf("%d", &n);

    // Input percentage scores
    printf("Enter the percentages of students:\n");
    for (int i = 0; i < n; i++) {
        printf("Student %d: ", i + 1);
        scanf("%f", &scores[i]);
    }

    int choice;
    printf("\nSelect Sorting Method:\n");
    printf("1. Selection Sort\n");
    printf("2. Bubble Sort\n");
    printf("Enter your choice (1 or 2): ");
    scanf("%d", &choice);

    // Sort and display top 5 scores based on the user's choice
    switch (choice) {
        case 1:
            selectionSort(scores, n);
            printf("\nSorted scores (Ascending Order):\n");
            for (int i = 0; i < n; i++) {
                printf("%.2f ", scores[i]);
            }
            printf("\n");
            displayTopFiveScores(scores, n);
            break;
            
        case 2:
            bubbleSort(scores, n);
            printf("\nSorted scores (Ascending Order):\n");
            for (int i = 0; i < n; i++) {
                printf("%.2f ", scores[i]);
            }
            printf("\n");
            displayTopFiveScores(scores, n);
            break;

        default:
            printf("Invalid choice!\n");
            break;
    }

    return 0;
}
